### PR TITLE
Add maxRetryCount capability, fix node v14 issue, fix capability constraints, fix unit tests

### PR DESCRIPTION
**Pull request by foad**
Apologies for squishing several things into one PR, if you want me to split it up into multiple PRs I will 👍 

 - Add maxRetryCount capability to override the default of 3 if desired
 - Fix issue on node v14 due to change in errno
 - Fix capability constraints not being correctly applied
 - Fix unit tests (most of them were failing)

### CODE REVIEW

File: lib/desired-caps.js [hunk 1]
```diff
@@ -21,6 +21,10 @@ const desiredCapConstraints = {
       'YouiEngine',
     ]
   },
+  deviceName: {
+    presence: true,
```
Comment (foad) [inline lib/desired-caps.js:25]: There was a unit test for this but the default capabilities allow it to be non-existent

File: lib/driver.js [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import { BaseDriver, DeviceSettings, errors } from 'appium-base-driver';
-import { desiredCapConstraints } from './desired-caps';
+import desiredCapConstraints from './desired-caps';
```
Comment (foad) [inline lib/driver.js:2]: `desired-cap.js` has a default export so these weren't being imported properly, so they were never actually being applied

File: lib/driver.js [hunk 3]
```diff
@@ -248,17 +250,16 @@ class YouiEngineDriver extends BaseDriver {
       return res;
     }
 
+    const platformName = caps.platformName.toLowerCase();
+
     // make sure that the capabilities has youiEngineAppAddress
-    if (!caps.youiEngineAppAddress) {
+    if (!caps.youiEngineAppAddress && platformName !== 'noproxy') {
```
Comment (foad) [inline lib/driver.js:256]: There was a unit test for this case but the logic didn't match, have changed the logic to match but if it's the unit test itself that's incorrect please let me know 🙏

File: lib/driver.js [hunk 4]
```diff
@@ -450,16 +451,14 @@ class YouiEngineDriver extends BaseDriver {
       retryCount++;
       connected = await connectedPromise;
 
-      if (!connected && errno === 'ECONNREFUSED') {
+      if (!connected && (errno === 'ECONNREFUSED' || errno === -61)) {
```
Comment (foad) [inline lib/driver.js:454]: On node v14+ errno is now a negative number representing the error itself (in this case -61 is ECONNREFUSED)

File: lib/driver.js [hunk 5]
```diff
@@ -450,16 +451,14 @@ class YouiEngineDriver extends BaseDriver {
...
         logger.debug('Connection refused, sleeping...');
         await sleep(2000);
         errno = 'EOK';
       }
 
-      if (!connected && retryCount === MAX_RETRY_COUNT - 1) {
-        logger.errorAndThrow(
-          'Failed to connect ' + MAX_RETRY_COUNT + ' times. Aborting.'
-        );
+      if (!connected && retryCount === this.maxRetryCount) {
```
Comment (foad) [inline lib/driver.js:460]: This happens after `retryCount++` so it was actually retrying one less time than expected, if you set maxRetryCount at 3 it'd only do 2 attempts

File: lib/driver.js (hunk 4 above)
Comment (lukeyoui) [reply lib/driver.js:454]: This good to know

File: lib/driver.js (hunk 5 above)
Comment (lukeyoui) [reply lib/driver.js:460]: Ironically I was just looking at this logic today as well. I'll keep it in your PR.

**Comment by lukeyoui** [comment]
Great stuff! I am merging in another two PR's today that I may need you to rebase on. After that I'll review again, and run through some of my tests to verify for regressions.

**Comment by foad** [comment]
@lukeyoui Thanks for the quick review, please let me know when I'm good to rebase and I'll do so 🙏

**Comment by lukeyoui** [comment]
@foad Good to rebase now! After which I'll verify again on my builds.

File: lib/driver.js [hunk 6]
```diff
@@ -393,11 +394,11 @@ class YouiEngineDriver extends BaseDriver {
         } else {
           PORT = 12345; //default port
         }
-        if(this.caps.platformName.toLowerCase() === 'android' && this.caps.youiEngineAppAddress.toLowerCase() === 'localhost'){
+        if (this.caps.platformName.toLowerCase() === 'android' && this.caps.youiEngineAppAddress.toLowerCase() === 'localhost') {
```
Comment (foad) [inline lib/driver.js:397]: Linter was complaining at me for this block

**Comment by foad** [comment]
@lukeyoui Rebased it now 👍

**Comment by lukeyoui** [comment]
Great! I'll start my tests, and I'll let you know if anything come up.

**Comment by lukeyoui** [comment]
Looks good to go! I'll merge it in and start working on a new driver version release.

**Review by lukeyoui — APPROVED** [review]

**Comment by foad** [comment]
Thanks Luke!

**Pull request merged**